Release/src/http/common/http_compression.cpp

```diff
@@ -0,0 +1,1110 @@
+        }
+
+        // Immediately try to instantiate a decompressor
+        token = encoding.substr(start, length);
+        auto d = web::http::compression::builtin::_make_decompressor(f, token);
+        if (d)
+        {
+            if (decompressor)
+            {
+                status_code code = status_codes::NotImplemented;
+                if (type == header_types::content_encoding)
+                {
+                    code = status_codes::UnsupportedMediaType;
+                }
+                throw http_exception(code, "Multiple compression algorithms not supported for a single request");
+            }
+
+            // We found our decompressor; store it off while we process the rest of the header
+            decompressor = std::move(d);
+        }
+        else
+        {
+            if (n != utility::string_t::npos)
+            {
+                if (type == header_types::transfer_encoding &&
+                    utility::details::str_iequal(_XPLATSTR("chunked"), token))
+                {
+                    throw http_exception(status_codes::BadRequest,
+                                         "Chunked must come last in the Transfer-Encoding header");
+                }
+            }
+            if (!decompressor && !f.empty() && (n != utility::string_t::npos || type == header_types::content_encoding))
+            {
+                // The first encoding type did not match; throw an informative
+                // exception with an encoding-type-appropriate HTTP error code
+                status_code code = status_codes::NotImplemented;
+                if (type == header_types::content_encoding)
+                {
+                    code = status_codes::UnsupportedMediaType;
+                }
+                throw http_exception(code, "Unsupported encoding type");
+            }
+        }
+    }
+
+    if (type == header_types::transfer_encoding && !utility::details::str_iequal(_XPLATSTR("chunked"), token))
+    {
+        throw http_exception(status_codes::BadRequest, "Transfer-Encoding header missing chunked");
+    }
+
+    // Either the response is compressed and we have a decompressor that can handle it, or
+    // built-in compression is not enabled and we don't have an alternate set of decompressors
+    return std::move(decompressor);
+}
+
+utility::string_t build_supported_header(header_types type,
+                                         const std::vector<std::shared_ptr<decompress_factory>>& factories)
+{
+    const std::vector<std::shared_ptr<decompress_factory>>& f =
+        factories.empty() ? web::http::compression::builtin::g_decompress_factories : factories;
+    utility::ostringstream_t os;
+    bool start;
+
+    _ASSERTE(type == header_types::te || type == header_types::accept_encoding);
+
+    // Add all specified algorithms and their weights to the header
+    start = true;
+    for (int i = 0; i < f.size(); i++)
```

Comment:
You'd have to ask the folks who wrote the compiler(s). :wink: You're right, though it's one of those things that I'd never noticed because it's always worked.

Your suggestion is cleaner, anyway.

Reply:
Maybe that's a /W4 /WX thing and cpprestsdk builds /W3 or something like that.